Unable to build successfully with HIP backends

When I try to include the flags `-D__OFFLOAD_HIP`, my compilation fails. The current build system is very confusing and the following are big concerns in my mind:
- Now we are including all HIP or CUDA backends by default and are having to exclude different backends if we choose to instead of following the general practice of choosing to include only the backends we want to have. This results in files not needed getting included in builds (see examples below).
- When I build for AMD GPUs, the build system seems to include the CUDA backend also? This could be a lack of understanding on my part, but this is causing problems.

Here is an example of building with `-D__NO_OFFLOAD_PW -D__OFFLOAD_HIP`. Note that `src/dbm/*.cu` and `src/grid/gpu/*.cu` files are being built by hipcc. I also see the `src/grid/hip/*` files being built with hipcc, but there is no problem there. I am able to exclude the dbm backend using `__NO_OFFLOAD_DBM` but the problem with the `grid/gpu` files remains.
```
[2022-10-12T00:08:26.459Z] hipcc -c -D__FFTW3 -D__parallel -D__COSMA -D__SCALAPACK -D__LIBXSMM -D__NO_OFFLOAD_PW -D__OFFLOAD_HIP -D__HIP_PLATFORM_AMD__ -I/opt/fftw/include  -D__LIBINT -D__MAX_CONTR=4 -D__LIBXC --amdgpu-target=gfx90a,gfx908,gfx906 -O3 -fdenormal-fp-math=ieee -fcuda-flush-denormals-to-zero -munsafe-fp-atomics -fPIC --std=c++11 /opt/cp2k/src/dbm/dbm_multiply_gpu.cu
[2022-10-12T00:08:26.962Z] hipcc -c -D__FFTW3 -D__parallel -D__COSMA -D__SCALAPACK -D__LIBXSMM -D__NO_OFFLOAD_PW -D__OFFLOAD_HIP -D__HIP_PLATFORM_AMD__ -I/opt/fftw/include  -D__LIBINT -D__MAX_CONTR=4 -D__LIBXC --amdgpu-target=gfx90a,gfx908,gfx906 -O3 -fdenormal-fp-math=ieee -fcuda-flush-denormals-to-zero -munsafe-fp-atomics -fPIC --std=c++11 /opt/cp2k/src/grid/gpu/grid_gpu_task_list.cu
[2022-10-12T00:08:27.213Z] [91mIn file included from /opt/cp2k/src/dbm/dbm_multiply_gpu.cu:15:
[2022-10-12T00:08:27.213Z] In file included from /opt/cp2k/src/dbm/dbm_multiply_gpu.h:15:
[2022-10-12T00:08:27.213Z] In file included from /opt/cp2k/src/dbm/dbm_multiply_internal.h:12:
[2022-10-12T00:08:27.213Z] /opt/cp2k/src/dbm/dbm_shard.h:10:10: fatal error: 'omp.h' file not found
[2022-10-12T00:08:27.213Z] #include <omp.h>
[2022-10-12T00:08:27.213Z]          ^~~~~~~
[2022-10-12T00:08:27.213Z] [0m[91m1 error generated when compiling for gfx906.
[2022-10-12T00:08:27.213Z] [0m[91mmake[3]: *** [/opt/cp2k/Makefile:521: dbm_multiply_gpu.o] Error 1
[2022-10-12T00:08:27.213Z] make[3]: *** Waiting for unfinished jobs....
[2022-10-12T00:08:27.465Z] [0m[91m/opt/cp2k/src/grid/gpu/grid_gpu_task_list.cu:12:10: fatal error: 'omp.h' file not found
[2022-10-12T00:08:27.465Z] #include <omp.h>
[2022-10-12T00:08:27.465Z]          ^~~~~~~
[2022-10-12T00:08:27.465Z] [0m[91m1 error generated when compiling for gfx906.
```
These `*.cu` files have some openmp pragmas for which we need to include the `-fopenmp` flag when compiling them. There is a problem with including `-fopenmp` in files compiled with `hipcc` because `hipcc` pulls in the openmp library found in llvm during linking. And since I use GNU compilers for everything else, we end up with `libomp.so` and `libgomp.so` in the final linked binary and performance drops.

If we can remove those openmp pragmas in these files or separate device code and host code into different translation units (compiled with different compilers), that would be one solution. I'd rather not build these files if they are not needed anyway.

--- a/src/dbm/dbm_multiply_gpu.c
+++ b/src/dbm/dbm_multiply_gpu.c
@@ -6,196 +6,16 @@
 /*----------------------------------------------------------------------------*/
 
 #include "../offload/offload_runtime.h"
-
 #if defined(__OFFLOAD) && !defined(__NO_OFFLOAD_DBM)
 
 #include "../offload/offload_library.h"
 #include "dbm_hyperparams.h"
 #include "dbm_mempool.h"
 #include "dbm_multiply_gpu.h"
+#include "dbm_multiply_gpu_kernel.h"
 
 #include <assert.h>
 #include <stdio.h>
-
-#define NUM_THREADS 256
-
-/*******************************************************************************
- * \brief Atomic add for doubles that also works prior to compute capability 6.
- * \author Ole Schuett
- ******************************************************************************/
-__device__ static void atomicAddDouble(double *address, double val) {
-  if (val == 0.0)
-    return;
-
-#if __CUDA_ARCH__ >= 600
-  atomicAdd(address, val); // part of gpu library
-#else
-  // https://docs.nvidia.com/gpu/gpu-c-programming-guide/index.html#atomic-functions
-  unsigned long long int *address_as_ull = (unsigned long long int *)address;
-  unsigned long long int old = *address_as_ull, assumed;
-
-  do {
-    assumed = old;
-    old = atomicCAS(address_as_ull, assumed,
-                    __double_as_longlong(val + __longlong_as_double(assumed)));
-
-    // Uses integer comparison to avoid hang in case of NaN (since NaN != NaN)
-  } while (assumed != old);
-
-#endif
-}
-
-/*******************************************************************************
- * \brief Returns the larger of two given integer (missing from the C standard)
- * \author Ole Schuett
- ******************************************************************************/
-__device__ constexpr static inline int imax(int x, int y) {
-  return (x > y ? x : y);
-}
-
-/*******************************************************************************
- * \brief Processes a task using tile dimensions give by template parameters.
- * \author Ole Schuett
- ******************************************************************************/
-template <int M, int N>
-__device__ static void process_task(const int m, const int n, const int k,
-                                    const double alpha, const double *block_a,
-                                    const double *block_b, double *block_c,
-                                    double *shared_mem) {
-
-  constexpr int K = NUM_THREADS / imax(M, N);
-
-  static_assert(K * imax(M, N) == NUM_THREADS, "Wasting threads.");
-  static_assert(M * N <= NUM_THREADS, "Not enough threads to cover tile.");
-
-  double *tile_a = shared_mem;
-  double *tile_b = &shared_mem[K * M];
-
-  // Layout threads to cover the entire tile.
-  // Indices x,y,z mark the position within the tile.
-  const int y = threadIdx.x / M;       // Can exceed N, guaranteed to reach K.
-  const int x = threadIdx.x - y * M;   // Fastest index, does not exceed M.
-  const int xT = threadIdx.x / N;      // Can exceed M, guaranteed to reach K.
-  const int yT = threadIdx.x - xT * N; // Fastest index, does not exceed N.
-
-  // Indices {ijl}_tile mark the beginning of the current tile.
-  for (int i_tile = 0; i_tile < m; i_tile += M) {
-    for (int j_tile = 0; j_tile < n; j_tile += N) {
-      double result = 0.0;
-      for (int l_tile = 0; l_tile < k; l_tile += K) {
-
-        // Load tile_a from global into shared memory.
-        if (x < M && y < K) {
-          const int i = i_tile + x;
-          const int l = l_tile + y;
-          const int idx = l * m + i; // transa = "N"
-          const bool load = (l < k && i < m);
-          tile_a[y * M + x] = (load) ? block_a[idx] : 0.0;
-        }
-
-        // Load tile_b from global into shared memory.
-        // Use transposed thread mapping to achieve coalesced memory reads.
-        if (yT < N && xT < K) {
-          const int j = j_tile + yT;
-          const int l = l_tile + xT;
-          const int idx = l * n + j; // transb = "T"
-          const bool load = (l < k && j < n);
-          tile_b[xT * N + yT] = (load) ? block_b[idx] : 0.0;
-        }
-
-        // Multiply tiles from shared memory.
-        __syncthreads();
-        if (x < M && y < N) {
-#pragma unroll
-          for (int z = 0; z < K; z++) {
-            result += tile_a[z * M + x] * tile_b[z * N + y];
-          }
-        }
-        __syncthreads();
-      }
-
-      // Add result tile to block_c in global memory.
-      if (x < M && y < N) {
-        const int i = i_tile + x;
-        const int j = j_tile + y;
-        if (i < m && j < n) {
-          const int idx = j * m + i;
-          // Need atomics because other thread blocks might work on same C.
-          atomicAddDouble(&block_c[idx], alpha * result);
-        }
-      }
-    }
-  }
-}
-
-/*******************************************************************************
- * \brief A generic matrix multiplication kernel.
- * \author Ole Schuett
- ******************************************************************************/
-__global__ static void process_batch_kernel(const double alpha,
-                                            const dbm_task_t *batch,
-                                            const double *pack_a_data,
-                                            const double *pack_b_data,
-                                            double *shard_c_data) {
-
-  __shared__ double shared_mem[2 * NUM_THREADS];
-
-  const dbm_task_t task = batch[blockIdx.x];
-
-  const int m = task.m;
-  const int n = task.n;
-  const int k = task.k;
-
-  const double *blk_a = &pack_a_data[task.offset_a];
-  const double *blk_b = &pack_b_data[task.offset_b];
-  double *blk_c = &shard_c_data[task.offset_c];
-
-  if (m <= 4 && n <= 4) {
-    process_task<4, 4>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
-
-  } else if (m <= 4 && n <= 8) {
-    process_task<4, 8>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
-
-  } else if (m <= 4 && n <= 16) {
-    process_task<4, 16>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
-
-  } else if (m <= 4 && n <= 32) {
-    process_task<4, 32>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
-
-  } else if (m <= 4) {
-    process_task<4, 64>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
-
-  } else if (m <= 8 && n <= 4) {
-    process_task<8, 4>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
-
-  } else if (m <= 16 && n <= 4) {
-    process_task<16, 4>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
-
-  } else if (m <= 32 && n <= 4) {
-    process_task<32, 4>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
-
-  } else if (n <= 4) {
-    process_task<64, 4>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
-
-  } else if (m <= 8 && n <= 8) {
-    process_task<8, 8>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
-
-  } else if (m <= 8 && n <= 16) {
-    process_task<8, 16>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
-
-  } else if (m <= 8) {
-    process_task<8, 32>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
-
-  } else if (m <= 16 && n <= 8) {
-    process_task<16, 8>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
-
-  } else if (n <= 8) {
-    process_task<32, 8>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
-
-  } else {
-    process_task<16, 16>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
-  }
-}
 
 /*******************************************************************************
  * \brief Internal routine for intializing the gpu backend.
@@ -325,13 +145,9 @@ void dbm_multiply_gpu_process_batch(const int ntasks, const dbm_task_t *batch,
   }
 
   // Launch kernel.
-  const int nblocks = ntasks; // TODO tune launch parameters.
-  const int threads_per_block = NUM_THREADS;
-  const size_t smem_per_block = 0;
-  process_batch_kernel<<<nblocks, threads_per_block, smem_per_block,
-                         shard_c_dev->stream>>>(
-      alpha, batch_dev, ctx->pack_a_dev.data, ctx->pack_b_dev.data,
-      shard_c_dev->data);
+  dbm_multiply_gpu_launch_kernel(shard_c_dev->stream, alpha, ntasks, batch_dev,
+                                 ctx->pack_a_dev.data, ctx->pack_b_dev.data,
+                                 shard_c_dev->data);
   OFFLOAD_CHECK(offloadGetLastError());
 
   // Wait for batch to be uploaded before refilling it.

--- a/src/dbm/dbm_multiply_gpu.h
+++ b/src/dbm/dbm_multiply_gpu.h
@@ -9,15 +9,10 @@
 #define DBM_MULTIPLY_GPU_H
 
 #include "../offload/offload_runtime.h"
-
 #if defined(__OFFLOAD) && !defined(__NO_OFFLOAD_DBM)
 
 #include "dbm_multiply_internal.h"
 #include "dbm_shard.h"
-
-#ifdef __cplusplus
-extern "C" {
-#endif
 
 /*******************************************************************************
  * \brief Internal struct for storing per shard gpu objects.
@@ -83,10 +78,6 @@ void dbm_multiply_gpu_download_results(dbm_multiply_gpu_context_t *ctx);
  * \author Ole Schuett
  ******************************************************************************/
 void dbm_multiply_gpu_stop(dbm_multiply_gpu_context_t *ctx);
-
-#ifdef __cplusplus
-}
-#endif
 
 #endif // defined(__OFFLOAD) && !defined(__NO_OFFLOAD_DBM)
 #endif

--- a/src/dbm/dbm_multiply_gpu_kernel.cu
+++ b/src/dbm/dbm_multiply_gpu_kernel.cu
@@ -1,0 +1,217 @@
+/*----------------------------------------------------------------------------*/
+/*  CP2K: A general program to perform molecular dynamics simulations         */
+/*  Copyright 2000-2022 CP2K developers group <https://cp2k.org>              */
+/*                                                                            */
+/*  SPDX-License-Identifier: BSD-3-Clause                                     */
+/*----------------------------------------------------------------------------*/
+
+#include "../offload/offload_runtime.h"
+#if defined(__OFFLOAD) && !defined(__NO_OFFLOAD_DBM)
+
+#include "dbm_multiply_gpu_kernel.h"
+
+#if defined(_OMP_H)
+#error "OpenMP should not be used in .cu files to accommodate HIP."
+#endif
+
+#define NUM_THREADS 256
+
+/*******************************************************************************
+ * \brief Atomic add for doubles that also works prior to compute capability 6.
+ * \author Ole Schuett
+ ******************************************************************************/
+__device__ static void atomicAddDouble(double *address, double val) {
+  if (val == 0.0)
+    return;
+
+#if __CUDA_ARCH__ >= 600
+  atomicAdd(address, val); // part of gpu library
+#else
+  // https://docs.nvidia.com/gpu/gpu-c-programming-guide/index.html#atomic-functions
+  unsigned long long int *address_as_ull = (unsigned long long int *)address;
+  unsigned long long int old = *address_as_ull, assumed;
+
+  do {
+    assumed = old;
+    old = atomicCAS(address_as_ull, assumed,
+                    __double_as_longlong(val + __longlong_as_double(assumed)));
+
+    // Uses integer comparison to avoid hang in case of NaN (since NaN != NaN)
+  } while (assumed != old);
+
+#endif
+}
+
+/*******************************************************************************
+ * \brief Returns the larger of two given integer (missing from the C standard)
+ * \author Ole Schuett
+ ******************************************************************************/
+__device__ constexpr static inline int imax(int x, int y) {
+  return (x > y ? x : y);
+}
+
+/*******************************************************************************
+ * \brief Processes a task using tile dimensions give by template parameters.
+ * \author Ole Schuett
+ ******************************************************************************/
+template <int M, int N>
+__device__ static void process_task(const int m, const int n, const int k,
+                                    const double alpha, const double *block_a,
+                                    const double *block_b, double *block_c,
+                                    double *shared_mem) {
+
+  constexpr int K = NUM_THREADS / imax(M, N);
+
+  static_assert(K * imax(M, N) == NUM_THREADS, "Wasting threads.");
+  static_assert(M * N <= NUM_THREADS, "Not enough threads to cover tile.");
+
+  double *tile_a = shared_mem;
+  double *tile_b = &shared_mem[K * M];
+
+  // Layout threads to cover the entire tile.
+  // Indices x,y,z mark the position within the tile.
+  const int y = threadIdx.x / M;       // Can exceed N, guaranteed to reach K.
+  const int x = threadIdx.x - y * M;   // Fastest index, does not exceed M.
+  const int xT = threadIdx.x / N;      // Can exceed M, guaranteed to reach K.
+  const int yT = threadIdx.x - xT * N; // Fastest index, does not exceed N.
+
+  // Indices {ijl}_tile mark the beginning of the current tile.
+  for (int i_tile = 0; i_tile < m; i_tile += M) {
+    for (int j_tile = 0; j_tile < n; j_tile += N) {
+      double result = 0.0;
+      for (int l_tile = 0; l_tile < k; l_tile += K) {
+
+        // Load tile_a from global into shared memory.
+        if (x < M && y < K) {
+          const int i = i_tile + x;
+          const int l = l_tile + y;
+          const int idx = l * m + i; // transa = "N"
+          const bool load = (l < k && i < m);
+          tile_a[y * M + x] = (load) ? block_a[idx] : 0.0;
+        }
+
+        // Load tile_b from global into shared memory.
+        // Use transposed thread mapping to achieve coalesced memory reads.
+        if (yT < N && xT < K) {
+          const int j = j_tile + yT;
+          const int l = l_tile + xT;
+          const int idx = l * n + j; // transb = "T"
+          const bool load = (l < k && j < n);
+          tile_b[xT * N + yT] = (load) ? block_b[idx] : 0.0;
+        }
+
+        // Multiply tiles from shared memory.
+        __syncthreads();
+        if (x < M && y < N) {
+#pragma unroll
+          for (int z = 0; z < K; z++) {
+            result += tile_a[z * M + x] * tile_b[z * N + y];
+          }
+        }
+        __syncthreads();
+      }
+
+      // Add result tile to block_c in global memory.
+      if (x < M && y < N) {
+        const int i = i_tile + x;
+        const int j = j_tile + y;
+        if (i < m && j < n) {
+          const int idx = j * m + i;
+          // Need atomics because other thread blocks might work on same C.
+          atomicAddDouble(&block_c[idx], alpha * result);
+        }
+      }
+    }
+  }
+}
+
+/*******************************************************************************
+ * \brief A generic matrix multiplication kernel.
+ * \author Ole Schuett
+ ******************************************************************************/
+__global__ static void process_batch_kernel(const double alpha,
+                                            const dbm_task_t *batch,
+                                            const double *pack_a_data,
+                                            const double *pack_b_data,
+                                            double *shard_c_data) {
+
+  __shared__ double shared_mem[2 * NUM_THREADS];
+
+  const dbm_task_t task = batch[blockIdx.x];
+
+  const int m = task.m;
+  const int n = task.n;
+  const int k = task.k;
+
+  const double *blk_a = &pack_a_data[task.offset_a];
+  const double *blk_b = &pack_b_data[task.offset_b];
+  double *blk_c = &shard_c_data[task.offset_c];
+
+  if (m <= 4 && n <= 4) {
+    process_task<4, 4>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
+
+  } else if (m <= 4 && n <= 8) {
+    process_task<4, 8>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
+
+  } else if (m <= 4 && n <= 16) {
+    process_task<4, 16>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
+
+  } else if (m <= 4 && n <= 32) {
+    process_task<4, 32>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
+
+  } else if (m <= 4) {
+    process_task<4, 64>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
+
+  } else if (m <= 8 && n <= 4) {
+    process_task<8, 4>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
+
+  } else if (m <= 16 && n <= 4) {
+    process_task<16, 4>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
+
+  } else if (m <= 32 && n <= 4) {
+    process_task<32, 4>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
+
+  } else if (n <= 4) {
+    process_task<64, 4>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
+
+  } else if (m <= 8 && n <= 8) {
+    process_task<8, 8>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
+
+  } else if (m <= 8 && n <= 16) {
+    process_task<8, 16>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
+
+  } else if (m <= 8) {
+    process_task<8, 32>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
+
+  } else if (m <= 16 && n <= 8) {
+    process_task<16, 8>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
+
+  } else if (n <= 8) {
+    process_task<32, 8>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
+
+  } else {
+    process_task<16, 16>(m, n, k, alpha, blk_a, blk_b, blk_c, shared_mem);
+  }
+}
+
+/*******************************************************************************
+ * \brief Internal routine for launching the GPU kernel.
+ *        All arguments are assumed to be device pointers.
+ * \author Ole Schuett
+ ******************************************************************************/
+void dbm_multiply_gpu_launch_kernel(const offloadStream_t stream,
+                                    const double alpha, const int ntasks,
+                                    const dbm_task_t *batch,
+                                    const double *pack_a_data,
+                                    const double *pack_b_data,
+                                    double *shard_c_data) {
+  const int nblocks = ntasks; // TODO tune launch parameters.
+  const int threads_per_block = NUM_THREADS;
+  const size_t smem_per_block = 0;
+  process_batch_kernel<<<nblocks, threads_per_block, smem_per_block, stream>>>(
+      alpha, batch, pack_a_data, pack_b_data, shard_c_data);
+}
+
+#endif // defined(__OFFLOAD) && !defined(__NO_OFFLOAD_DBM)
+
+// EOF

--- a/src/dbm/dbm_multiply_gpu_kernel.h
+++ b/src/dbm/dbm_multiply_gpu_kernel.h
@@ -1,0 +1,39 @@
+/*----------------------------------------------------------------------------*/
+/*  CP2K: A general program to perform molecular dynamics simulations         */
+/*  Copyright 2000-2022 CP2K developers group <https://cp2k.org>              */
+/*                                                                            */
+/*  SPDX-License-Identifier: BSD-3-Clause                                     */
+/*----------------------------------------------------------------------------*/
+
+#ifndef DBM_MULTIPLY_GPU_KERNEL_H
+#define DBM_MULTIPLY_GPU_KERNEL_H
+
+#include "../offload/offload_runtime.h"
+#if defined(__OFFLOAD) && !defined(__NO_OFFLOAD_DBM)
+
+#include "dbm_multiply_internal.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*******************************************************************************
+ * \brief Internal routine for launching the GPU kernel.
+ *        All arguments are assumed to be device pointers.
+ * \author Ole Schuett
+ ******************************************************************************/
+void dbm_multiply_gpu_launch_kernel(const offloadStream_t stream,
+                                    const double alpha, const int ntasks,
+                                    const dbm_task_t *batch,
+                                    const double *pack_a_data,
+                                    const double *pack_b_data,
+                                    double *shard_c_data);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif // defined(__OFFLOAD) && !defined(__NO_OFFLOAD_DBM)
+#endif
+
+// EOF

--- a/src/dbm/dbm_multiply_internal.h
+++ b/src/dbm/dbm_multiply_internal.h
@@ -8,9 +8,6 @@
 #ifndef DBM_MULTIPLY_INTERNAL_H
 #define DBM_MULTIPLY_INTERNAL_H
 
-#include "dbm_hyperparams.h"
-#include "dbm_shard.h"
-
 /*******************************************************************************
  * \brief Internal struct for storing a dbm_block_t plus its norm.
  * \author Ole Schuett

--- a/src/grid/gpu/grid_gpu_collocate.cu
+++ b/src/grid/gpu/grid_gpu_collocate.cu
@@ -22,6 +22,10 @@
 #include "grid_gpu_collint.h"
 #include "grid_gpu_collocate.h"
 
+#if defined(_OMP_H)
+#error "OpenMP should not be used in .cu files to accommodate HIP."
+#endif
+
 /*******************************************************************************
  * \brief Collocate a single grid point with distance d{xyz} from center.
  * \author Ole Schuett

--- a/src/grid/gpu/grid_gpu_integrate.cu
+++ b/src/grid/gpu/grid_gpu_integrate.cu
@@ -22,6 +22,10 @@
 #include "grid_gpu_collint.h"
 #include "grid_gpu_integrate.h"
 
+#if defined(_OMP_H)
+#error "OpenMP should not be used in .cu files to accommodate HIP."
+#endif
+
 // Teen registers are sufficient to integrate lp <= 2 with a single grid sweep.
 #define GRID_N_CXYZ_REGISTERS 10
 

--- a/src/grid/gpu/grid_gpu_task_list.cu
+++ b/src/grid/gpu/grid_gpu_task_list.cu
@@ -9,7 +9,6 @@
 #if defined(__OFFLOAD) && !defined(__NO_OFFLOAD_GRID)
 
 #include <assert.h>
-#include <omp.h>
 #include <stdint.h>
 #include <stdio.h>
 #include <stdlib.h>
@@ -22,6 +21,10 @@
 #include "grid_gpu_collocate.h"
 #include "grid_gpu_integrate.h"
 #include "grid_gpu_task_list.h"
+
+#if defined(_OMP_H)
+#error "OpenMP should not be used in .cu files to accommodate HIP."
+#endif
 
 /*******************************************************************************
  * \brief Create a single task and precompute as much as possible.

--- a/src/grid/hip/grid_hip_collocate.cu
+++ b/src/grid/hip/grid_hip_collocate.cu
@@ -25,6 +25,10 @@
 #include "grid_hip_internal_header.h"
 #include "grid_hip_prepare_pab.h"
 
+#if defined(_OMP_H)
+#error "OpenMP should not be used in .cu files to accommodate HIP."
+#endif
+
 namespace rocm_backend {
 /*******************************************************************************
  * \brief Decontracts the subblock, going from spherical to cartesian harmonics.

--- a/src/grid/hip/grid_hip_context.cu
+++ b/src/grid/hip/grid_hip_context.cu
@@ -31,6 +31,10 @@ extern "C" {
 
 #include "grid_hip_task_list.h"
 
+#if defined(_OMP_H)
+#error "OpenMP should not be used in .cu files to accommodate HIP."
+#endif
+
 /*******************************************************************************
  * \brief Allocates a task list for the GPU backend.
  *        See grid_ctx.h for details.

--- a/src/grid/hip/grid_hip_integrate.cu
+++ b/src/grid/hip/grid_hip_integrate.cu
@@ -35,6 +35,10 @@
 namespace cg = cooperative_groups;
 #endif
 
+#if defined(_OMP_H)
+#error "OpenMP should not be used in .cu files to accommodate HIP."
+#endif
+
 namespace rocm_backend {
 // do a warp reduction and return the final sum to thread_id = 0
 template <typename T>

--- a/src/pw/gpu/pw_gpu_kernels.cu
+++ b/src/pw/gpu/pw_gpu_kernels.cu
@@ -5,6 +5,9 @@
 /*  SPDX-License-Identifier: GPL-2.0-or-later                                 */
 /*----------------------------------------------------------------------------*/
 
+#include "../../offload/offload_runtime.h"
+#if defined(__OFFLOAD) && !defined(__NO_OFFLOAD_PW)
+
 #include "pw_gpu_kernels.h"
 
 #if defined(_OMP_H)
@@ -128,5 +131,7 @@ void pw_gpu_launch_scatter(double *c, const double *pwcc, const double scale,
   pw_scatter<<<numBlocks, threadsPerBlock, 0, stream>>>(c, pwcc, scale, ngpts,
                                                         nmaps, ghatmap);
 }
+
+#endif // defined(__OFFLOAD) && !defined(__NO_OFFLOAD_PW)
 
 // EOF

--- a/src/pw/gpu/pw_gpu_kernels.cu
+++ b/src/pw/gpu/pw_gpu_kernels.cu
@@ -7,6 +7,10 @@
 
 #include "pw_gpu_kernels.h"
 
+#if defined(_OMP_H)
+#error "OpenMP should not be used in .cu files to accommodate HIP."
+#endif
+
 /*******************************************************************************
  * \brief   Performs a out-of-place copy of a double precision vector (first
  *          half filled) into a double precision complex vector on the GPU.

--- a/src/pw/gpu/pw_gpu_kernels.h
+++ b/src/pw/gpu/pw_gpu_kernels.h
@@ -9,6 +9,7 @@
 #define PW_GPU_KERNELS_H
 
 #include "../../offload/offload_runtime.h"
+#if defined(__OFFLOAD) && !defined(__NO_OFFLOAD_PW)
 
 #ifdef __cplusplus
 extern "C" {
@@ -48,6 +49,7 @@ void pw_gpu_launch_scatter(double *c, const double *pwcc, const double scale,
 }
 #endif
 
+#endif // defined(__OFFLOAD) && !defined(__NO_OFFLOAD_PW)
 #endif
 
 // EOF

--- a/tools/precommit/check_file_properties.py
+++ b/tools/precommit/check_file_properties.py
@@ -153,6 +153,9 @@ def check_file(path: pathlib.Path) -> typing.List[str]:
     if fn_ext not in (".pot", ".patch") and basefn != "Makefile" and "\t" in content:
         warnings += [f"{path}: contains tab character"]
 
+    if fn_ext == ".cu" and "#if defined(_OMP_H)\n#error" not in content:
+        warnings += [f"{path}: misses check against OpenMP usage"]
+
     # check banner
     year = datetime.utcnow().year
     bsd_licensed = any(str(path).startswith(d) for d in BSD_DIRECTORIES)
@@ -197,6 +200,8 @@ def check_file(path: pathlib.Path) -> typing.List[str]:
     flags = {flag for flag in flags if not FLAG_EXCEPTIONS_RE.match(flag)}
 
     for flag in sorted(flags):
+        if flag == "_OMP_H" and fn_ext == ".cu":
+            continue
         if flag not in get_install_txt():
             warnings += [f"{path}: Flag '{flag}' not mentioned in INSTALL.md"]
         if flag not in get_flags_src():

